Reenable slow gradcheck CI and skip composite compliance tests

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85849
* #79056
* __->__ #88093

Fixes: https://github.com/pytorch/pytorch/issues/88010

~~Composite compliance tests don't use gradcheck, but can be slow: https://raw.githubusercontent.com/pytorch/test-infra/generated-stats/stats/slow-tests.json. The cc tests for pca/svd low-rank combine to increase test time by >30min, so we may want to skip these test for normal CI as well.~~

~~CC forward AD tests can be slow because one is testing all subclass + non-subclass combination for both primal and tangents, e.g. if my function has 3 arguments, I end up running forward AD 2^3 x 2^3 = 64 times.~~

The above might not actually matter because it actually applies to the other slow-gradcheck shard (and indeed its test time has been reduced by 30 minutes). The real issue is that the automatic test sharding doesn't work well when some tests that don't have test time information:

dynamo/test_recompile_ux
inductor/test_smoke
lazy/test_bindings
lazy/test_meta_kernel
test_comparison_utils
test_hub
doctest

test_ops_gradients alone takes 3.3h, but it is in the same shard as the tests above.

Solution 1: just manually shard again:
- this might be the way forward if having tests that don't have timing information is something that we can't avoid
- it just makes the call run_test slightly less clean

Solution 2: allow list instead of block list tests that should run under slow:
- this is good because we won't have to periodically skip slow tests anymore, but runs the risk of accidentally skipping some tests that use gradcheck

Solution 3: skip the tests above (because they don't use gradcheck)
- pros: no risk of skipping tests that use gradcheck
- cons: hard to skip doctest in a super clean way + there might be more tests that don't having timing information, so this might be annoying to do 

The current PR applies solution 3 to minimize the blast radius, solution 1 seems OK also if there are just too many tests that lack timing information, solution 2 may require more thorough investigation to confirm which tests use gradcheck (so I'm less inclined to do that)

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

soulitzer committed:
Skip composite compliant tests on slow gradcheck CI

[ghstack-poisoned]

zou3519 commented on test/test_ops.py:
```diff
@@ -1333,6 +1333,7 @@ def _tensor_requires_grad(x):
         self.fail(msg)
 
 
+@skipIfSlowGradcheckEnv
 class TestCompositeCompliance(TestCase):
```
Why do we run tests other than slow gradcheck tests during slow gradcheck CI?

soulitzer replied:
> 1. Why do we run tests other than slow gradcheck tests during slow gradcheck CI?

Yup, completely agree here. Tests that use gradcheck are spread across different test classes, so there's some trickiness here though. As a next step, it might be a good idea to have an allow list of the ones that we know use gradcheck. I originally hesitated because I didn't have an exhaustive list, and I only looked at the test classes that take the longest to run.

2. TestCompositeCompliance is being skipped here because it doesn't use gradcheck at all

albanD reviewed (approved):
SGTM if that makes it fast enough

soulitzer committed:
Update on "Reenable slow gradcheck CI and skip composite compliance tests"


Fixes: https://github.com/pytorch/pytorch/issues/88010

Composite compliance tests don't use gradcheck, but can be slow: https://raw.githubusercontent.com/pytorch/test-infra/generated-stats/stats/slow-tests.json. The cc tests for pca/svd low-rank combine to increase test time by >30min, so we may want to skip these test for normal CI as well.

CC forward AD tests can be slow because one is testing all subclass + non-subclass combination for both primal and tangents, e.g. if my function has 3 arguments, I end up running forward AD 2^3 x 2^3 = 64 times.

[ghstack-poisoned]

soulitzer commented:
This has been subsumed by https://github.com/pytorch/pytorch/pull/88216
